Firestore/core/test/unit/api/load_bundle_task_test.cc

```diff
@@ -0,0 +1,415 @@
+        task.RemoveObserver(handle1);
+
+        handle2 = task.ObserveState(
+            LoadBundleTaskState::InProgress, [&](LoadBundleTaskProgress p) {
+              (void)p;
+              queue.push(2);
+              // handle3
+              task.ObserveState(LoadBundleTaskState::InProgress,
+                                [&](LoadBundleTaskProgress p) {
+                                  (void)p;
+                                  queue.push(3);
+                                  task.RemoveObserver(handle2);
+                                });
+            });
+      });
+
+  // Running handle1, which registers handle2 and removes itself
+  task.UpdateProgress(SuccessProgress());
+  EXPECT_EQ(1, queue.pop());
+
+  // Running handle2, which registers handle3
+  task.UpdateProgress(SuccessProgress());
+  EXPECT_EQ(2, queue.pop());
+
+  // Running handle2 and handle3. handle2 registers another handle3, then
+  // handle3 removes handle2.
+  task.UpdateProgress(SuccessProgress());
+  EXPECT_EQ(2, queue.pop());
+  EXPECT_EQ(3, queue.pop());
+
+  // Running two handle3
+  task.UpdateProgress(SuccessProgress());
+  EXPECT_EQ(3, queue.pop());
+  EXPECT_EQ(3, queue.pop());
+  EXPECT_TRUE(queue.empty());
+}
+
+TEST(LoadBundleTaskTest, ProgressObservesUntilSuccess) {
+  LoadBundleTask task(CreateUserQueue());
+  BlockingQueue<LoadBundleTaskProgress> queue;
+  task.ObserveState(LoadBundleTaskState::InProgress,
+                    [&](LoadBundleTaskProgress p) { queue.push(p); });
+
+  task.UpdateProgress(InitialProgress());
+  EXPECT_EQ(InitialProgress(), queue.pop());
+
+  task.UpdateProgress(Progress(2, 5));
+  EXPECT_EQ(Progress(2, 5), queue.pop());
+
+  task.SetSuccess(SuccessProgress());
+  EXPECT_EQ(SuccessProgress(), queue.pop());
+
+  EXPECT_TRUE(queue.empty());
+}
+
+TEST(LoadBundleTaskTest, ProgressObservesUntilError) {
+  LoadBundleTask task(CreateUserQueue());
+  BlockingQueue<LoadBundleTaskProgress> queue;
+  task.ObserveState(LoadBundleTaskState::InProgress,
+                    [&](LoadBundleTaskProgress p) { queue.push(p); });
+
+  task.UpdateProgress(InitialProgress());
+  EXPECT_EQ(InitialProgress(), queue.pop());
+
+  task.UpdateProgress(Progress(2, 5));
+  EXPECT_EQ(Progress(2, 5), queue.pop());
+
+  task.SetError();
+  auto expected = Progress(2, 5);
+  expected.set_state(LoadBundleTaskState::Error);
+  EXPECT_EQ(expected, queue.pop());
+
+  EXPECT_TRUE(queue.empty());
+}
+
+TEST(LoadBundleTaskTest, ProgressObservesInitialError) {
+  LoadBundleTask task(CreateUserQueue());
+  BlockingQueue<LoadBundleTaskProgress> queue;
+  task.ObserveState(LoadBundleTaskState::InProgress,
+                    [&](LoadBundleTaskProgress p) { queue.push(p); });
+
+  task.SetError();
+  auto expected = LoadBundleTaskProgress();
+  expected.set_state(LoadBundleTaskState::Error);
+  EXPECT_EQ(expected, queue.pop());
+
+  EXPECT_TRUE(queue.empty());
+}
+
+TEST(LoadBundleTaskTest, NoObserversAlsoWork) {
+  LoadBundleTask task(CreateUserQueue());
```

Comment:
It looks like every test creates the `task` in the same manner. Consider adding a test class and making `task` its member variable.

Reply:
Done.